impeller/typographer/backends/skia/typographer_context_skia.cc

```diff
@@ -402,43 +405,68 @@ static Rect ComputeGlyphSize(const SkFont& font,
                         scaled_bounds.fBottom);
 };
 
-static void CollectNewGlyphs(const std::shared_ptr<GlyphAtlas>& atlas,
-                             const FontGlyphMap& font_glyph_map,
-                             std::vector<FontGlyphPair>& new_glyphs,
-                             std::vector<Rect>& glyph_sizes) {
-  for (const auto& font_value : font_glyph_map) {
-    const ScaledFont& scaled_font = font_value.first;
-    const FontGlyphAtlas* font_glyph_atlas =
-        atlas->GetFontGlyphAtlas(scaled_font.font, scaled_font.scale);
-
-    auto metrics = scaled_font.font.GetMetrics();
-
-    SkFont sk_font(
-        TypefaceSkia::Cast(*scaled_font.font.GetTypeface()).GetSkiaTypeface(),
-        metrics.point_size, metrics.scaleX, metrics.skewX);
-    sk_font.setEdging(SkFont::Edging::kAntiAlias);
-    sk_font.setHinting(SkFontHinting::kSlight);
-    sk_font.setEmbolden(metrics.embolden);
-    // Rather than computing the bounds at the requested point size and scaling
-    // up the bounds, we scale up the font size and request the bounds. This
-    // seems to give more accurate bounds information.
-    sk_font.setSize(sk_font.getSize() * scaled_font.scale);
-    sk_font.setSubpixel(true);
-
-    if (font_glyph_atlas) {
-      for (const SubpixelGlyph& glyph : font_value.second) {
-        if (!font_glyph_atlas->FindGlyphBounds(glyph)) {
-          new_glyphs.emplace_back(scaled_font, glyph);
-          glyph_sizes.push_back(
-              ComputeGlyphSize(sk_font, glyph, scaled_font.scale));
+void TypographerContextSkia::CollectNewGlyphs(
+    const std::shared_ptr<GlyphAtlas>& atlas,
+    const std::vector<std::shared_ptr<TextFrame>>& text_frames,
+    std::vector<FontGlyphPair>& new_glyphs,
+    std::vector<Rect>& glyph_sizes) {
+  for (const auto& frame : text_frames) {
+    // TODO(jonahwilliams): unless we destroy the atlas (which we know about),
```

Comment:
Leaving this as the next optimizations. With a little plumbing when we recreate the atlas we can actually avoid any hashing on frames with no new text.

Reply:
sgtm